text/0000-subslice-pattern-syntax.md

```diff
@@ -0,0 +1,207 @@
+
+ When used to match against a non-reference slice (`[u8]`), `x @ ..` would attempt to bind
+by-value, which would fail in the case that users haven't enabled `feature(unsized_locals)`
+(since otherwise it's not possible to bind `[u8]` to a variable directly).
+
+`..`/`IDENT @ ..` is not a full pattern syntax, but rather a part of slice, tuple and tuple
+struct pattern syntaxes. In particular, `..` is not accepted by the `pat` macro matcher.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None known.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+More complex syntaxes derived from `..` are possible, they use additional tokens to avoid the
+ambiguity with ranges, for example
+[`..PAT..`](https://github.com/rust-lang/rust/issues/23121#issuecomment-301485132), or
+[`.. @ PAT`](https://github.com/rust-lang/rust/issues/23121#issuecomment-280920062) or
+[`PAT @ ..`](https://github.com/rust-lang/rust/issues/23121#issuecomment-280906823), or other
+similar alternatives.  
+We reject these syntaxes because they only bring benefits in incredibly contrived cases using a
+feature that doesn't even exist yet, but normally they only add symbolic noise.
+
+More radical syntax changes not keeping consistency with `..`, for example
+[`[1, 2, 3, 4] ++ ref v`](https://github.com/rust-lang/rust/issues/23121#issuecomment-289220169).
+
+### `..PAT` or `PAT..`
+
+If `..` is used in the meaning "match the subslice (`>=0` elements) and ignore it", then it's
+reasonable to expect that syntax for "match the subslice to a pattern" should be some variation
+on `..`.  
+The two simplest variations are `..PAT` and `PAT..`.
+
+#### Ambiguity
+
+The issue is that these syntaxes are ambiguous with half-bounded ranges `..END` and `BEGIN..`.  
+To be precise, such ranges are not currently supported in patterns, but they may be supported in
+the future.
+
+Syntactic ambiguity is not inherently bad. We see it every day in expressions like
+`a + b * c`. What is important is to disambiguate it reasonably by default and have a way to
+group operands in the alternative way when default disambiguation turns out to be incorrect.  
+In case of slice patterns the subslice interpretation seems overwhelmingly more likely, so we
+can take it as a default.  
+There was no visible demand for implementing half-bounded ranges in patterns so far, but if they
+are implemented in the future they will be able to be used in slice patterns as well, but they
+will require explicit grouping with recently implemented
+[parentheses in patterns](https://github.com/rust-lang/rust/pull/48500) (`[a, (..end)]`) or an
+explicitly written start boundary (`[a, 0 .. end]`).  
+We can also make *some* disambiguation effort and, for example, interpret `..LITERAL` as a
+range because `LITERAL` can never match a subslice. Time will show if such an effort is necessary
+or not.
+
+If/when half-bounded ranges are supported in patterns, for better future compatibility we'll need
```

Comment:
```suggestion
If/when half-bounded ranges are supported in patterns, for better future compatibility we could decide
```
As above, I don't think we should make such commitments since it seems rather ad-hoc.